src/parquet/column/scanner.h

```diff
@@ -0,0 +1,206 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_COLUMN_SCANNER_H
+#define PARQUET_COLUMN_SCANNER_H
+
+#include <memory>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include "parquet/column/reader.h"
+#include "parquet/thrift/parquet_types.h"
+
+namespace parquet_cpp {
+
+static constexpr size_t DEFAULT_SCANNER_BATCH_SIZE = 128;
+
+class Scanner {
+ public:
+  explicit Scanner(std::shared_ptr<ColumnReader> reader,
+      size_t batch_size = DEFAULT_SCANNER_BATCH_SIZE) :
+      reader_(reader),
+      batch_size_(batch_size),
+      level_offset_(0),
+      levels_buffered_(0),
+      value_offset_(0),
+      values_buffered_(0) {
+    // TODO: don't allocate for required fields
+    def_levels_.resize(batch_size_);
+    rep_levels_.resize(batch_size_);
+  }
+
+  static std::shared_ptr<Scanner> Make(std::shared_ptr<ColumnReader> col_reader,
+      size_t batch_size = DEFAULT_SCANNER_BATCH_SIZE);
+
+  virtual void PrintNext(std::ostream& out, int width) = 0;
```

Comment:
classes with virtual functions need a virtual dtor. Our compile flags should warn on this.